### PR TITLE
[Transforms] Add `NodePropertySplit` transform

### PR DESCRIPTION
**Pull request by gvbazhenov**
New functional for creating node-level data splits, which allows to induce distributional shifts in graph and conduct experiments with graph models in more challenging setups. It is implemented via the `NodePropertySplit` transform and allows to use the class itself or its static method `mask_nodes_by_property` (the latter is for users who already knows their structural node property and just need to perform data splitting).

Test coverage is provided and passed successfully. I decided to edit `transforms.rst` in `docs/source/modules` to make the static method `mask_nodes_by_property` visible in documentation, but not sure if this is the right way.

### CODE REVIEW

**Comment by gvbazhenov** [comment]
I see that `networkx` is not a necessary dependency for `torch-geometric`. How can I fix the implementation so that the library and documentation build successfully?

**Comment by rusty1s** [comment]
You can move the imports of `networkx` to the corresponding functions that make use of it. In this case, you probably need to also replace the `nx.Graph` type hints with something more general such as `Any`.

**Comment by gvbazhenov** [comment]
As I can see, test fails on `import networkx as nx` when `NodePropertySplit` instance calls `to_networkx`. This is rather strange, since I just rely on the existing utilities in `utils/convert`. Could you please explain me what I am doing wrong?

File: test/transforms/test_node_property_split.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+from itertools import combinations
+
+import numpy as np
+import torch
+
+from torch_geometric.datasets import graph_generator
+from torch_geometric.transforms import NodePropertySplit
+
+
+def test_node_property_split():
```
Comment (EdisonLeeeee) [inline test/transforms/test_node_property_split.py:11]: You might need to add a decorator `@withPackage('networkx')` here to add the `networkx` package during tests.

Comment (gvbazhenov) [reply test/transforms/test_node_property_split.py:13]: It works, thank you!

File: torch_geometric/transforms/node_property_split.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    :obj:`"out_valid_mask"`, :obj:`"out_test_mask"`.
+
+    This class implements 3 particular strategies for inducing
+    distributional shifts in graph — based on
+    **popularity**, **locality** or **density**.
+
+    Args:
+        property_name (str): The name of the node property to be used,
+            which must be ``'popularity'``, ``'locality'`` or ``'density'``.
+
```
Comment (EdisonLeeeee) [inline torch_geometric/transforms/node_property_split.py:39]: Remove the blank lines from the code.

File: torch_geometric/transforms/node_property_split.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    :obj:`"in_test_mask"`, and the remaining 2 — for the OOD nodes:
+    :obj:`"out_valid_mask"`, :obj:`"out_test_mask"`.
+
+    This class implements 3 particular strategies for inducing
+    distributional shifts in graph — based on
+    **popularity**, **locality** or **density**.
+
+    Args:
+        property_name (str): The name of the node property to be used,
+            which must be ``'popularity'``, ``'locality'`` or ``'density'``.
```
Comment (EdisonLeeeee) [inline torch_geometric/transforms/node_property_split.py:38]:
```suggestion
            which must be :obj:`'popularity'`, :obj:`'locality'` or :obj:`'density'`.
```

File: torch_geometric/transforms/node_property_split.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+        property_name = "popularity"
+        part_ratios = [0.3, 0.1, 0.1, 0.3, 0.2]
+        tranaform = NodePropertySplit(property_name, part_ratios)
+
+        data = transform(data)
+    """
+    def __init__(
+        self,
+        property_name: str,
+        part_ratios: list,
```
Comment (EdisonLeeeee) [inline torch_geometric/transforms/node_property_split.py:73]:
```suggestion
        part_ratios: List[float],
```

File: torch_geometric/transforms/node_property_split.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+        ascending: Optional[bool] = True,
```
Comment (EdisonLeeeee) [inline torch_geometric/transforms/node_property_split.py:74]:
```suggestion
        ascending: bool = True,
```

File: torch_geometric/transforms/node_property_split.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        random_seed: Optional[int] = None,
+    ):
+        assert property_name in [
+            "popularity",
+            "locality",
+            "density",
+        ], "`property_name` has to be 'popularity', 'locality', or 'density'"
+
+        assert sum(part_ratios) == 1.0, "`part_ratios` must sum to 1.0"
```
Comment (EdisonLeeeee) [inline torch_geometric/transforms/node_property_split.py:83]: Add an assert statement to verify that the length of `part_ratios` is equal to 5.

File: torch_geometric/transforms/node_property_split.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+        self.property_name = property_name
+        self.part_ratios = part_ratios
+        self.ascending = ascending
+        self.random_seed = random_seed
+
+    def __repr__(self) -> str:
```
Comment (EdisonLeeeee) [inline torch_geometric/transforms/node_property_split.py:90]: Move this section of the code to the bottom.

**Review by EdisonLeeeee — APPROVED** [review]
LGTM. Thanks for the PR!

**Comment by gvbazhenov** [comment]
@EdisonLeeeee @rusty1s thanks for your PR review and help!

**Comment by gvbazhenov** [comment]
@rusty1s @EdisonLeeeee excuse me, could you please share when it will be possible to merge this pull request? Is there anything else that requires my participation?

**Comment by rusty1s** [comment]
Will take a final look soon, but overall this looks good. Sorry for the delay, I was quite busy with PyG 2.3 release :(

**Comment by gvbazhenov** [comment]
> Will take a final look soon, but overall this looks good. Sorry for the delay, I was quite busy with PyG 2.3 release :(

Roger that, thank you!

**Review by rusty1s — APPROVED** [review]
Thank you :)

**Pull request merged**